File: platform/macos/Bridge/NativeRemoteLayout.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import TidyVNC

extension tidyvnc_desktop_layout: ABIValue {}
extension tidyvnc_desktop_layout_request: ABIValue {}

public struct NativeRemoteScreen: Equatable, Sendable {
  public let id: UInt32, x: UInt32, y: UInt32, width: UInt32, height: UInt32, flags: UInt32
  public init(id: UInt32, x: UInt32, y: UInt32, width: UInt32, height: UInt32, flags: UInt32 = 0) {
    self.id = id; self.x = x; self.y = y; self.width = width; self.height = height; self.flags = flags
  }
  var abiValue: tidyvnc_remote_screen { .init(id:id,x:x,y:y,width:width,height:height,flags:flags) }
}

public struct NativeRemoteLayout: Equatable, Sendable {
  public let width: UInt32, height: UInt32
  public let screens: [NativeRemoteScreen]
  public init(width: UInt32, height: UInt32, screens: [NativeRemoteScreen]) throws {
    guard screens.count <= 255 else { throw NativeError(.invalidArgument,"Too many remote screens") }
    self.width = width; self.height = height; self.screens = screens
    _ = try withABI { value in try checked { tidyvnc_desktop_layout_validate(value,$0) } }
  }
  func withABI<T>(_ body: (UnsafePointer<tidyvnc_desktop_layout_request>) throws -> T) rethrows -> T {
    try screens.map(\.abiValue).withUnsafeBufferPointer { screens in
      var request = abi(tidyvnc_desktop_layout_request.self)
      request.width = width; request.height = height; request.screen_count = UInt32(screens.count)
      request.screens = screens.baseAddress
      return try withUnsafePointer(to:&request,body)
    }
  }
}

public struct NativeRemoteDesktop: Sendable {
  public let layout: NativeRemoteLayout
  public let snapshot: NativeSnapshot
  init(_ value: tidyvnc_desktop_layout) throws {
    guard value.screen_count > 0 && value.screen_count <= 255 else { throw NativeError(.internalFailure,"Invalid remote layout") }
    snapshot = NativeSnapshot(value.snapshot)
    let screens = withUnsafeBytes(of:value.screens) { bytes in
      bytes.bindMemory(to:tidyvnc_remote_screen.self).prefix(Int(value.screen_count)).map {
        NativeRemoteScreen(id:$0.id,x:$0.x,y:$0.y,width:$0.width,height:$0.height,flags:$0.flags)
      }
    }
    layout = try NativeRemoteLayout(width:value.snapshot.width,height:value.snapshot.height,screens:screens)
  }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeRemoteLayout.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

public readonly record struct NativeRemoteScreen(uint Id, uint X, uint Y, uint Width, uint Height, uint Flags = 0)
{
    internal tidyvnc_remote_screen ToAbi() => new() { id = Id, x = X, y = Y, width = Width, height = Height, flags = Flags };
}

/// <summary>A validated remote desktop layout request (NativeRemoteLayout).</summary>
public sealed record NativeRemoteLayout
{
    public uint Width { get; }
    public uint Height { get; }
    public IReadOnlyList<NativeRemoteScreen> Screens { get; }

    public unsafe NativeRemoteLayout(uint width, uint height, IReadOnlyList<NativeRemoteScreen> screens)
    {
        if (screens.Count > 255) throw new NativeError(NativeStatus.InvalidArgument, "Too many remote screens");
        Width = width; Height = height; Screens = screens.ToArray();
        var values = Screens.Select(s => s.ToAbi()).ToArray();
        var error = Abi.Init<tidyvnc_error>();
        fixed (tidyvnc_remote_screen* p = values)
        {
            var request = Abi.Init<tidyvnc_desktop_layout_request>();
            request.width = width; request.height = height; request.screen_count = (uint)values.Length; request.screens = p;
            Abi.Check(NativeMethods.tidyvnc_desktop_layout_validate(&request, &error), &error);
        }
    }

    public bool Equals(NativeRemoteLayout? other)
        => other is not null && Width == other.Width && Height == other.Height && Screens.SequenceEqual(other.Screens);

    public override int GetHashCode() => HashCode.Combine(Width, Height, Screens.Count);
}

public sealed record NativeRemoteDesktop(NativeRemoteLayout Layout, NativeSnapshot Snapshot)
{
    internal static NativeRemoteDesktop From(tidyvnc_desktop_layout value)
    {
        if (value.screen_count == 0 || value.screen_count > 255) throw new NativeError(NativeStatus.InternalFailure, "Invalid remote layout");
        var screens = new List<NativeRemoteScreen>();
        for (var i = 0; i < value.screen_count; i++)
        {
            var s = value.screens[i];
            screens.Add(new NativeRemoteScreen(s.id, s.x, s.y, s.width, s.height, s.flags));
        }
        return new NativeRemoteDesktop(new NativeRemoteLayout(value.snapshot.width, value.snapshot.height, screens),
                                       NativeSnapshot.From(value.snapshot));
    }
}
